### PR TITLE
Regional/Zonal instance group manager.

### LINKED ISSUE #59: GCP: Managed Instance Groups
> An instance group is a collection of virtual machine (VM) instances that you can manage as a single entity.

> Compute Engine offers two kinds of VM instance groups, managed and unmanaged:

> Managed instance groups (MIGs) let you operate apps on multiple identical VMs. You can make your workloads scalable and highly available by taking advantage of automated MIG services, including: autoscaling, autohealing, regional (multiple zone) deployment, and automatic updating.

This uses the Instance Group Manager API: https://cloud.google.com/compute/docs/reference/rest/v1/instanceGroupManagers

This work should also include support for the Region Instance Group Manager API:
https://cloud.google.com/compute/docs/reference/rest/v1/regionInstanceGroupManagers

image

### PR DESCRIPTION
**Pull request by sshyub**
Fixes #59 

Base branch must be changed to `master` once #82 is merged.

### CODE REVIEW

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerUpdatePolicy.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+import com.google.api.services.compute.model.InstanceGroupManagerUpdatePolicy;
+import gyro.core.resource.Diffable;
+import gyro.core.validation.ValidStrings;
+import gyro.google.Copyable;
+
+public class ComputeInstanceGroupManagerUpdatePolicy extends Diffable
+    implements Copyable<InstanceGroupManagerUpdatePolicy> {
+
+    private String instanceRedistributionType;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeInstanceGroupManagerUpdatePolicy.java]: The attribute is unused and should be removed.

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatus.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import com.google.api.services.compute.model.InstanceGroupManagerStatus;
+import gyro.core.resource.Diffable;
+import gyro.core.resource.Output;
+import gyro.google.Copyable;
+
+public class ComputeInstanceGroupManagerStatus extends Diffable implements Copyable<InstanceGroupManagerStatus> {
+
+    private Boolean isStable;
+
+    private ComputeInstanceGroupManagerStatusVersionTarget versionTarget;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatus.java]: Should be removed if not present in the sdk.

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+
+    private String fingerprint;
+
+    private InstanceTemplateResource instanceTemplate;
+
+    private List<ComputeNamedPort> namedPort;
+
+    // TODO: target pool resource
+    // https://github.com/perfectsense/gyro-google-provider/issues/79
+    //    private List<AbstractTargetPoolResource> targetPools;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]: Should be Removed and be done part of a separate pr when target pool is available

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+    }
+
+    public void setInstanceGroupLink(String instanceGroupLink) {
+        this.instanceGroupLink = instanceGroupLink;
+    }
+
+    /**
+     * The URL for this managed instance group. The server defines this URL.
+     */
+    @Output
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]:
```suggestion
    @Id
    @Output
```

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        waitForCompletion(client, operation);
+        refresh();
+    }
+
+    @Override
+    public void doUpdate(GyroUI ui, State state, Resource current, Set<String> changedFieldNames)
+        throws Exception {
+        // TODO: target pool resource
+        // https://github.com/perfectsense/gyro-google-provider/issues/79
+        //    private List<AbstractTargetPoolResource> targetPools;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java]: Should be removed and be done as a separate PR when target pool is available

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        setDistributionPolicy(Optional.ofNullable(model.getDistributionPolicy())
+            .map(e -> {
+                ComputeDistributionPolicy computeDistributionPolicy = newSubresource(
+                    ComputeDistributionPolicy.class);
+                computeDistributionPolicy.copyFrom(e);
+                return computeDistributionPolicy;
+            })
+            .orElse(null));
+        // Do NOT update region with a full url as this should be a name.
+        //        setRegion(model.getRegion());
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java]: Shouldn't this be fine now as the `@Immutable` annotation is present ??

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+    private void setInstanceTemplate() throws Exception {
+        Compute client = createClient(Compute.class);
+        RegionInstanceGroupManagersSetTemplateRequest request = new RegionInstanceGroupManagersSetTemplateRequest();
+        request.setInstanceTemplate(getInstanceTemplate().getSelfLink());
+        Operation operation = client.regionInstanceGroupManagers()
+            .setInstanceTemplate(getProjectId(), getRegion(), getName(), request)
+            .execute();
+        waitForCompletion(client, operation);
+        refresh();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java]: Refresh should be called at the end the create/update unless you need to reset a value before.

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    private void patch(InstanceGroupManager instanceGroupManager) throws Exception {
+        Compute client = createClient(Compute.class);
+        Operation operation = client.regionInstanceGroupManagers()
+            .patch(getProjectId(), getRegion(), getName(), instanceGroupManager)
+            .execute();
+        waitForCompletion(client, operation);
+        refresh();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java]: Refresh should be called at the end the create/update unless you need to reset a value before.

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+                    autoHealingPolicies = diffableAutoHealingPolicy
+                        .stream()
+                        .map(ComputeInstanceGroupManagerAutoHealingPolicy::copyTo)
+                        .collect(Collectors.toList());
+                }
+                instanceGroupManager.setAutoHealingPolicies(autoHealingPolicies);
+                shouldPatch = true;
+            } else if (changedFieldName.equals("named-port")) {
+                // TODO: investigate as GCP UI allows updating of named ports.
+            }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java]: If the api does not allowing the update of named ports then this should not be present.

File: src/main/java/gyro/google/compute/InstanceGroupManagerResource.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+                    autoHealingPolicies = diffableAutoHealingPolicy
+                        .stream()
+                        .map(ComputeInstanceGroupManagerAutoHealingPolicy::copyTo)
+                        .collect(Collectors.toList());
+                }
+                instanceGroupManager.setAutoHealingPolicies(autoHealingPolicies);
+                shouldPatch = true;
+            } else if (changedFieldName.equals("named-port")) {
+                // TODO: investigate as GCP UI allows updating of named ports.
+            }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerResource.java]: If the api does not allowing the update of named ports then this should not be present.

File: src/main/java/gyro/google/compute/InstanceGroupManagerResource.java [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        waitForCompletion(client, operation);
+        refresh();
+    }
+
+    @Override
+    public void doUpdate(GyroUI ui, State state, Resource current, Set<String> changedFieldNames)
+        throws Exception {
+        // TODO: target pool resource
+        // https://github.com/perfectsense/gyro-google-provider/issues/79
+        //    private List<AbstractTargetPoolResource> targetPools;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerResource.java]: Should be removed and be done as a separate PR when target pool is available

File: src/main/java/gyro/google/compute/InstanceGroupManagerResource.java [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+            .execute();
+        waitForCompletion(client, operation);
+    }
+
+    @Override
+    public void copyFrom(InstanceGroupManager model) {
+        super.copyFrom(model);
+
+        // Do NOT update zone with a full url as this should be a name.
+        //        setZone(model.getZone());
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerResource.java]: Shouldn't this be fine now as the @Immutable annotation is present ??

File: src/main/java/gyro/google/compute/InstanceGroupManagerResource.java [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        refresh();
+    }
+
+    private void patch(InstanceGroupManager instanceGroupManager) throws Exception {
+        Compute client = createClient(Compute.class);
+        Operation operation = client.instanceGroupManagers()
+            .patch(getProjectId(), getZone(), getName(), instanceGroupManager)
+            .execute();
+        waitForCompletion(client, operation);
+        refresh();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerResource.java]: Refresh should be called at the end the create/update unless you need to reset a value before.

File: src/main/java/gyro/google/compute/InstanceGroupManagerResource.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+    private void setInstanceTemplate() throws Exception {
+        Compute client = createClient(Compute.class);
+        InstanceGroupManagersSetInstanceTemplateRequest request = new InstanceGroupManagersSetInstanceTemplateRequest();
+        request.setInstanceTemplate(getInstanceTemplate().getSelfLink());
+        Operation operation = client.instanceGroupManagers()
+            .setInstanceTemplate(getProjectId(), getZone(), getName(), request)
+            .execute();
+        waitForCompletion(client, operation);
+        refresh();
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerResource.java]: Refresh should be called at the end the create/update unless you need to reset a value before.

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+ * Example
+ * -------
+ *
+ * .. code-block:: gyro
+ *
+ *    region-instance-group-manager: $(external-query google::compute-region-instance-group-manager { name: 'region-instance-group-manager-example' })
+ */
+@Type("compute-region-instance-group-manager")
+public class RegionInstanceGroupManagerFinder
+    extends GoogleFinder<Compute, InstanceGroupManager, InstanceGroupManagerResource> {
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java]:
```suggestion
    extends GoogleFinder<Compute, InstanceGroupManager, RegionInstanceGroupManagerResource> {
```

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java [hunk 16]
```diff
@@ -0,0 +1,446 @@
...
+    //        if (targetPools == null) {
+    //            targetPools = new ArrayList<>();
+    //        }
+    //
+    //        return targetPools;
+    //    }
+    //
+    //    public void setTargetPools(List<AbstractTargetPoolResource> targetPools) {
+    //        this.targetPools = targetPools;
+    //    }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]: Should be removed and be done as a separate PR when target pool is available

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java [hunk 17]
```diff
@@ -0,0 +1,446 @@
...
+        //        List<AbstractTargetPoolResource> diffableTargetPools = null;
+        //        List<String> targetPools = model.getTargetPools();
+        //
+        //        if (targetPools != null && !targetPools.isEmpty()) {
+        //            diffableTargetPools = targetPools
+        //                .stream()
+        //                .map(e -> findById(AbstractTargetPoolResource.class, e))
+        //                .collect(Collectors.toList());
+        //        }
+        //        setTargetPools(diffableTargetPools);
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]: Should be removed and be done as a separate PR when target pool is available

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java [hunk 18]
```diff
@@ -0,0 +1,446 @@
...
+            .map(InstanceTemplateResource::getSelfLink)
+            .ifPresent(instanceGroupManager::setInstanceTemplate);
+        instanceGroupManager.setNamedPorts(getNamedPort().stream()
+            .map(ComputeNamedPort::copyTo)
+            .collect(Collectors.toList()));
+        // TODO: https://github.com/perfectsense/gyro-google-provider/issues/79
+        //        instanceGroupManager.setTargetPools(getTargetPools()
+        //            .stream()
+        //            .map(AbstractTargetPoolResource::getSelfLink)
+        //            .collect(Collectors.toList()));
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]: Should be removed and be done as a separate PR when target pool is available

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java (hunk 3 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]: Yeah, it should be handled in a separate PR, but better to leave the comment as it will be helpful when implementing it.

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java (hunk 5 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java]: Yeah, it should be handled in a separate PR, but better to leave the comment as it will be helpful when implementing it.

File: src/main/java/gyro/google/compute/InstanceGroupManagerResource.java (hunk 11 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/InstanceGroupManagerResource.java]: Yeah, it should be handled in a separate PR, but better to leave the comment as it will be helpful when implementing it.

File: src/main/java/gyro/google/compute/InstanceGroupManagerResource.java (hunk 12 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/InstanceGroupManagerResource.java]: Yeah, it should be handled in a separate PR, but better to leave the comment as it will be helpful when implementing it.

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java (hunk 16 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]: Yeah, it should be handled in a separate PR, but better to leave the comment as it will be helpful when implementing it.

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java (hunk 17 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]: Yeah, it should be handled in a separate PR, but better to leave the comment as it will be helpful when implementing it.

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java (hunk 18 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]: Yeah, it should be handled in a separate PR, but better to leave the comment as it will be helpful when implementing it.

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java (hunk 6 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/RegionInstanceGroupManagerResource.java]: GCP doesn't accept a full url in the request.

File: src/main/java/gyro/google/compute/InstanceGroupManagerResource.java (hunk 12 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/InstanceGroupManagerResource.java]: GCP doesn't accept a full url in the request.

File: src/main/java/gyro/google/compute/ComputeDistributionPolicyZoneConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+        DistributionPolicyZoneConfiguration distributionPolicyZoneConfiguration = new DistributionPolicyZoneConfiguration();
+        distributionPolicyZoneConfiguration.setZone(getZoneLink());
+        return distributionPolicyZoneConfiguration;
+    }
+
+    @Override
+    public void copyFrom(DistributionPolicyZoneConfiguration model) {
+        setZoneLink(model.getZone());
+    }
+
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeDistributionPolicyZoneConfiguration.java]: For consistency, it would be good to accept the zone from the user, and then convert it to the zone url.
As per the example it is `projects/<project-id>/zones/<zone-name>` which is simple to construct.

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+    extends GoogleFinder<Compute, InstanceGroupManager, RegionInstanceGroupManagerResource> {
+
+    private String name;
+
+    private String region;
+
+    /**
+     * User assigned name for the instance group manager.
+     */
+    @Required
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java]: It should not be required when calling the `findAllGoogle()` method
```suggestion
```

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    /**
+     * The zone where the managed instance group is located (for zonal resources).
+     */
+    @Required
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java]: It should not be required when calling the findAllGoogle() method
```suggestion
```

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+            List<InstanceGroupManager> items = response.getItems();
+
+            if (items == null) {
+                break;
+            }
+            allInstanceGroupManagers.addAll(items);
+            nextPageToken = response.getNextPageToken();
+            request.setPageToken(nextPageToken);
+        } while (nextPageToken != null);
+        return allInstanceGroupManagers;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java]: The `findAllGoogle()` should not use parameters.
In this case `client.instanceGroupManagers().aggregatedList()` should be used to get all instance groups and then filtered based on which has the region attribute filled classifying as regional instance group.

See https://github.com/perfectsense/gyro-google-provider/blob/master/src/main/java/gyro/google/compute/RegionDiskFinder.java#L74-L92 for reference.

File: src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+    extends GoogleFinder<Compute, InstanceGroupManager, InstanceGroupManagerResource> {
+
+    private String name;
+
+    private String zone;
+
+    /**
+     * User assigned name for the instance group manager.
+     */
+    @Required
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java]: It should not be required when calling the findAllGoogle() method
```suggestion
```

File: src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    /**
+     * The zone where the managed instance group is located (for zonal resources).
+     */
+    @Required
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java]: It should not be required when calling the findAllGoogle() method
```suggestion
```

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    /**
+     * The zone where the managed instance group is located (for zonal resources).
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java]:
```suggestion
     * The region where the managed instance group is located.
```

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+@Type("compute-region-instance-group-manager")
+public class RegionInstanceGroupManagerFinder
+    extends GoogleFinder<Compute, InstanceGroupManager, RegionInstanceGroupManagerResource> {
+
+    private String name;
+
+    private String region;
+
+    /**
+     * User assigned name for the instance group manager.
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java]: Because this is the doc for the external query, user assigned is not appropriate
```suggestion
     * Name for the instance group manager.
```

File: src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java [hunk 27]
```diff
@@ -0,0 +1,100 @@
...
+            request.setPageToken(nextPageToken);
+        } while (nextPageToken != null);
+        return allInstanceGroupManagers;
+    }
+
+    @Override
+    protected List<InstanceGroupManager> findGoogle(Compute client, Map<String, String> filters) throws Exception {
+        return Collections.singletonList(client.regionInstanceGroupManagers()
+            .get(getProjectId(), filters.get("region"), filters.get("name"))
+            .execute());
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/RegionInstanceGroupManagerFinder.java]: Should handle the case when only region is provided - list all the managed instances for that region.

File: src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+@Type("compute-instance-group-manager")
+public class InstanceGroupManagerFinder
+    extends GoogleFinder<Compute, InstanceGroupManager, InstanceGroupManagerResource> {
+
+    private String name;
+
+    private String zone;
+
+    /**
+     * User assigned name for the instance group manager.
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java]: Because this is the doc for the external query, user assigned is not appropriate
```suggestion
     * Name for the instance group manager.
```

File: src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    /**
+     * The zone where the managed instance group is located (for zonal resources).
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java]:
```suggestion
     * The zone where the managed instance group is located.
```

File: src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+            request.setPageToken(nextPageToken);
+        } while (nextPageToken != null);
+        return allInstanceGroupManagers;
+    }
+
+    @Override
+    protected List<InstanceGroupManager> findGoogle(Compute client, Map<String, String> filters) throws Exception {
+        return Collections.singletonList(client.instanceGroupManagers()
+            .get(getProjectId(), filters.get("zone"), filters.get("name"))
+            .execute());
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java]: Should handle the case when only zone is provided - list all the managed instances for that zone.

File: src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java [hunk 31]
```diff
@@ -0,0 +1,99 @@
...
+            List<InstanceGroupManager> items = response.getItems();
+
+            if (items == null) {
+                break;
+            }
+            allInstanceGroupManagers.addAll(items);
+            nextPageToken = response.getNextPageToken();
+            request.setPageToken(nextPageToken);
+        } while (nextPageToken != null);
+        return allInstanceGroupManagers;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/InstanceGroupManagerFinder.java]: The findAllGoogle() should not use parameters.
In this case client.instanceGroupManagers().aggregatedList() should be used to get all instance groups and then filtered based on which has the zone attribute filled classifying as zonal instance group.

See https://github.com/perfectsense/gyro-google-provider/blob/master/src/main/java/gyro/google/compute/RegionDiskFinder.java#L74-L92 for reference.

File: src/main/java/gyro/google/compute/ComputeDistributionPolicy.java [hunk 32]
```diff
@@ -0,0 +1,86 @@
...
+
+    public DistributionPolicy copyTo() {
+        DistributionPolicy distributionPolicy = new DistributionPolicy();
+        List<ComputeDistributionPolicyZoneConfiguration> zone = getZone();
+
+        if (!zone.isEmpty()) {
+            distributionPolicy.setZones(zone.stream()
+                .map(ComputeDistributionPolicyZoneConfiguration::copyTo)
+                .collect(Collectors.toList()));
+        }
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeDistributionPolicy.java]: As zone is a required field, isn't the check redundant ?
we could directly set it
```
 distributionPolicy.setZones(zone.stream()
    .map(ComputeDistributionPolicyZoneConfiguration::copyTo)
    .collect(Collectors.toList()));
```

File: src/main/java/gyro/google/compute/ComputeDistributionPolicy.java [hunk 33]
```diff
@@ -0,0 +1,86 @@
...
+                .stream()
+                .map(zone -> {
+                    ComputeDistributionPolicyZoneConfiguration computeDistributionPolicyZoneConfiguration = newSubresource(
+                        ComputeDistributionPolicyZoneConfiguration.class);
+                    computeDistributionPolicyZoneConfiguration.copyFrom(zone);
+                    return computeDistributionPolicyZoneConfiguration;
+                })
+                .collect(Collectors.toList());
+        }
+        setZone(diffableZone);
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeDistributionPolicy.java]: As zones is a required field do we expect it to be ever null or empty ?

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java [hunk 34]
```diff
@@ -0,0 +1,497 @@
...
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    /**
+     * The target number of running instances for this managed instance group.
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]:
```suggestion
     * The target number of running instances for this managed instance group. Defaults to ``0``.
```

File: src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java [hunk 35]
```diff
@@ -0,0 +1,497 @@
...
+        this.targetSize = targetSize;
+    }
+
+    /**
+     * The autohealing policy for this managed instance group.
+     * You can specify only one value.
+     *
+     * @subresource gyro.google.compute.ComputeInstanceGroupManagerAutoHealingPolicy
+     */
+    @Updatable
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/AbstractInstanceGroupManagerResource.java]:
```suggestion
    @CollectionMax(1)
    @Updatable
```

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerUpdatePolicy.java [hunk 36]
```diff
@@ -0,0 +1,146 @@
...
+import com.google.api.services.compute.model.InstanceGroupManagerUpdatePolicy;
+import gyro.core.resource.Diffable;
+import gyro.core.validation.ValidStrings;
+import gyro.google.Copyable;
+
+public class ComputeInstanceGroupManagerUpdatePolicy extends Diffable
+    implements Copyable<InstanceGroupManagerUpdatePolicy> {
+
+    // instanceRedistributionType is not available from Google SDK yet.
+    //    private String instanceRedistributionType;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeInstanceGroupManagerUpdatePolicy.java]: As it is not supported so should be removed.

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerVersion.java [hunk 37]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    /**
+     * Specifies the intended number of instances to be created from the instanceTemplate.
+     * The final number of instances created from the template will be equal to:
+     *
+     * - If expressed as a fixed number, the minimum of either ``target-size.fixed`` or ``instance-group-manager.target-size`` is used.
+     * - if expressed as a percent, the ``target-size`` would be (``target-size.percent``/100 * ``instance-group-manager.target-size``) If there is a remainder, the number is rounded up.
+     *
+     * If unset, this version will update any remaining instances not updated by another version. Read Starting a canary update for more information.
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeInstanceGroupManagerVersion.java]: Either `Read Starting a canary update for more information.` remove this portion or provide a link to this, otherwise its . a bit arbitrary.

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatusVersionTarget.java [hunk 38]
```diff
@@ -0,0 +1,43 @@
...
+        return isReached;
+    }
+
+    public void setIsReached(Boolean isReached) {
+        this.isReached = isReached;
+    }
+
+    @Override
+    public String primaryKey() {
+        // TODO: implement
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatusVersionTarget.java]: what more to implement ?

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatusVersionTarget.java [hunk 39]
```diff
@@ -0,0 +1,43 @@
...
+        return "";
+    }
+}
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatusVersionTarget.java]: This class should be removed as is not used.

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatus.java [hunk 40]
```diff
@@ -0,0 +1,53 @@
...
+import gyro.core.resource.Diffable;
+import gyro.core.resource.Output;
+import gyro.google.Copyable;
+
+public class ComputeInstanceGroupManagerStatus extends Diffable implements Copyable<InstanceGroupManagerStatus> {
+
+    private Boolean isStable;
+
+    // versionTarget is not available from Google SDK yet.
+    //    private ComputeInstanceGroupManagerStatusVersionTarget versionTarget;
```
Comment (deepanjan90) [inline src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatus.java]: Should be removed.

File: src/main/java/gyro/google/compute/ComputeDistributionPolicy.java (hunk 32 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/ComputeDistributionPolicy.java]: Annotation doesn't guarantee the value is not `null`.

File: src/main/java/gyro/google/compute/ComputeDistributionPolicy.java (hunk 33 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/ComputeDistributionPolicy.java]: Annotation doesn't guarantee the value is not `null`.

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerUpdatePolicy.java (hunk 36 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/ComputeInstanceGroupManagerUpdatePolicy.java]: This is to make clear that why the field wasn't implemented.

File: src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatus.java (hunk 40 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/ComputeInstanceGroupManagerStatus.java]: This is to make clear that why the field wasn't implemented.

File: src/main/java/gyro/google/compute/ComputeDistributionPolicyZoneConfiguration.java (hunk 19 above)
Comment (sshyub) [reply src/main/java/gyro/google/compute/ComputeDistributionPolicyZoneConfiguration.java]: We can't get `projectid` from `Diffable`, so I'll leave it as it is for now.
